### PR TITLE
JS: add a "../" removing taint-step for js/path-injection

### PR DESCRIPTION
**Pull request by erik-krogh**
Gets us a TN for the fix in CVE-2017-16107

### CODE REVIEW

File: javascript/ql/src/semmle/javascript/security/dataflow/TaintedPathCustomizations.qll [hunk 1]
```diff
@@ -239,6 +240,57 @@ module TaintedPath {
...
+      exists(RegExpLiteral literal, RegExpTerm term |
+        getArgument(0).getALocalSource().asExpr() = literal and
+        (term instanceof RegExpStar or term instanceof RegExpPlus) and
+        term.getChild(0) = getADotDotSlashMatcher()
+      |
+        literal.getRoot() = term
+        or
+        exists(RegExpSequence seq | seq.getNumChild() = 2 and literal.getRoot() = seq |
+          seq.getChild(0) instanceof RegExpCaret and
+          seq.getChild(1) = term
```
[the remark this answers is not in the document]
Comment (erik-krogh) [reply javascript/ql/src/semmle/javascript/security/dataflow/TaintedPathCustomizations.qll:263]: The replace call does not work on `./.././` in the first place (anchor or not), it only works on normalized paths.

And e.g. `path.normalize("./.././../foo") === "../../foo"`.  

So I don't see why the anchor should be mandatory.

**Comment by erik-krogh** [comment]
A [smoke test evaluation](https://git.semmle.com/erik/dist-compare-reports/tree/profiling-erik-krogh.northeurope.cloudapp.azure.com_1585904714031) looks fine.

File: javascript/ql/src/semmle/javascript/security/dataflow/TaintedPathCustomizations.qll (hunk 1 above)
[the remark this answers is not in the document]
Comment (asgerf) [reply javascript/ql/src/semmle/javascript/security/dataflow/TaintedPathCustomizations.qll:263]: Good point 👍

File: javascript/ql/src/semmle/javascript/security/dataflow/TaintedPath.qll [hunk 2]
```diff
@@ -210,7 +210,7 @@ module TaintedPath {
       exists(DotDotSlashPrefixRemovingReplace call |
         src = call.getInput() and
         dst = call.getOutput() and
-        dstlabel.isAbsolute() and // result can be absolute
+        (srclabel.isNonNormalized() or dstlabel.isAbsolute()) and // if src is normalized, then dst must be absolute (if dst is relative, then dst is sanitized)
```
Comment (asgerf) [inline javascript/ql/src/semmle/javascript/security/dataflow/TaintedPath.qll:213]: Ok now it's making my head explode 😕

Let's just go over the whole table:
- Normalized relative -> empty set (completely sanitized)
- Raw relative ->
  - Raw relative. Example: `foo/../bar/../..` becomes `foo/bar/../..`
  - Raw absolute. Example: `..//foo/bar/../..` becomes `/foo/bar/../..`
  - Normalized absolute. Example: `..//foo/bar` becomes `/foo/bar`
  - Normalized relative. Example: `..././foo/bar` becomes `../foo/bar`
- Normalized absolute ->
  - Normalized absolute (can't contain `../` so unaffected)
- Raw absolute ->
  - Raw absolute. Example: `/foo/../bar/../..` becomes `/foo/bar/../..`
  - Normalized absolute. Example: `/foo/../bar/` becomes `/foo/bar`.

The line below here doesn't preserve raw->normalized transitions. Missing the `raw relative/absolute -> normalized absolute` in particular can cause FNs for the `FsPathSinkWithoutUpwardNavigation` sink.

Perhaps I'm being a stickler but I'd like it if we can get this transition table exactly right.

Comment (erik-krogh) [reply javascript/ql/src/semmle/javascript/security/dataflow/TaintedPath.qll:213]: > Ok now it's making my head explode confused

I also had some trouble conceptualizing what was supposed to happen in each case, and how to express this as compact QL. (I considered basically writing out the entire truth table in QL). 

> Perhaps I'm being a stickler but I'd like it if we can get this transition table exactly right.

It's fine. Thanks for writing out the table.

File: javascript/ql/src/semmle/javascript/security/dataflow/TaintedPathCustomizations.qll [hunk 3]
```diff
@@ -239,6 +240,57 @@ module TaintedPath {
...
+    DataFlow::Node input;
+    DataFlow::Node output;
+
+    DotDotSlashPrefixRemovingReplace() {
+      this.getCalleeName() = "replace" and
+      input = getReceiver() and
+      output = this and
+      exists(RegExpLiteral literal, RegExpTerm term |
+        getArgument(0).getALocalSource().asExpr() = literal and
+        (term instanceof RegExpStar or term instanceof RegExpPlus) and
```
Comment (esbena) [inline javascript/ql/src/semmle/javascript/security/dataflow/TaintedPathCustomizations.qll:256]: Minor comment post-merge: this motivates a reuse of `SuperlinearBackTracking::InfiniteRepetitionQuantifier`, but it probably won' matter in this case.

File: javascript/ql/src/semmle/javascript/security/dataflow/TaintedPathCustomizations.qll [hunk 4]
```diff
@@ -239,6 +240,57 @@ module TaintedPath {
...
+  /**
+   * Gets a RegExpTerm that matches a variation of "../".
+   */
+  private RegExpTerm getADotDotSlashMatcher() {
+    result.getAMatchedString() = "../"
+    or
+    exists(RegExpSequence seq | seq = result |
+      seq.getChild(0).getConstantValue() = "." and
+      seq.getChild(1).getConstantValue() = "." and
+      seq.getAChild().getAMatchedString() = "/"
```
Comment (esbena) [inline javascript/ql/src/semmle/javascript/security/dataflow/TaintedPathCustomizations.qll:288]: Minor comment post-merge: should this have been `seq.getAChild(2)`?

File: javascript/ql/src/semmle/javascript/security/dataflow/TaintedPathCustomizations.qll (hunk 3 above)
Comment (erik-krogh) [reply javascript/ql/src/semmle/javascript/security/dataflow/TaintedPathCustomizations.qll:256]: This case is not quite the same as `InfiniteRepetitionQuantifier`.   
Here we need the lower-bound to be zero or one occurrences, and `InfiniteRepetitionQuantifier` doesn't care about the lower-bound.  
(And would e.g. flag `/(?:\.\·\/){2,}/`).   

So I think we can skip such a reuse for now.

**Pull request merged**